library/src/main/java/io/ona/kujaku/views/KujakuMapView.java

```diff
@@ -660,8 +698,59 @@ private void checkPermissions() {
     }
 
     @Override
-    public void showCurrentLocationBtn(boolean isVisible) {
-        currentLocationBtn.setVisibility(isVisible ? VISIBLE : GONE);
+    public void addFeaturePoints(FeatureCollection featureCollection) throws JSONException {
+        JSONArray featuresArray = this.featureCollectionJSON.getJSONArray("features");
+        for (com.mapbox.geojson.Feature feature : featureCollection.features()) {
+            String featureId = feature.id();
+            if (!featureMap.containsKey(featureId)) {
+                featureMap.put(featureId, featuresArray.length());
+                featuresArray.put(new JSONObject(feature.toJson()));
+            }
+        }
+        if (mapboxMap != null) {
+            ((GeoJsonSource) mapboxMap.getSource(primaryGeoJsonSource.getId())).setGeoJson(featureCollectionJSON.toString());
+        }
+    }
+
+    @Override
+    public void updateFeaturePointProperties(FeatureCollection featureCollection) throws JSONException {
+        List<com.mapbox.geojson.Feature> newFeatures = new ArrayList<>();
+        JSONArray featuresArray = this.featureCollectionJSON.getJSONArray("features");
+        for (com.mapbox.geojson.Feature feature : featureCollection.features()) {
+            String featureId = feature.id();
+            if (featureMap.containsKey(featureId)) {
+                int featureIndex = featureMap.get(featureId);
+                featuresArray.put(featureIndex, new JSONObject(feature.toJson()));
+            } else {
+                newFeatures.add(feature);
+            }
+        }
+        // add new features if any
+        FeatureCollection newFeatureCollection = FeatureCollection.fromFeatures(newFeatures);
+        addFeaturePoints(newFeatureCollection);
+        if (mapboxMap != null) {
+            ((GeoJsonSource) mapboxMap.getSource(primaryGeoJsonSource.getId())).setGeoJson(featureCollectionJSON.toString());
+        }
+    }
+
+    public MapboxMap getMapBoxMap() {
+        return this.mapboxMap;
+    }
+
+    public void initializePrimaryGeoJsonSource(String sourceId) {
+        primaryGeoJsonSource = new GeoJsonSource(sourceId, featureCollectionJSON.toString());
```

Comment:
I'm thinking you meant layer and not style? Is there a straightforward way to extract the source from a layer?

Also, what data format spec do you have in mind when extracting the source from the layer because I think these two are usually separate.

Reply:
Its possible for styles to defined a datasource id in the style, and the style defines specifically targets objects that defined in the datasource.

e.g in reveal this is how we initialize the datasource defined in the style
` GeoJsonSource geoJsonSource = mapboxMap.getSourceAs("reveal-data-set");`